brave/src/main/java/com/linecorp/armeria/client/brave/BraveClient.java

```diff
@@ -52,6 +56,16 @@
 
     private static final Logger logger = LoggerFactory.getLogger(BraveClient.class);
 
+    private static final Scope CLIENT_REQUEST_DECORATING_SCOPE = new Scope() {
+        @Override
+        public void close() {}
```

Comment:
Could you explain why we need to decorate a scope with a no-op scope like this?

Reply:
The following code decorates the given scope and returns `decoratedScope` by wrapping the `scope`. 
```java
Scope decoratedScope = currentTraceContext.decorateScope(span.context(), scope)
```

While decorating the scope, it pushes some contexts such as MDC to are stored to a thread-local.
The activated contexts will be removed when `decoratedScope.close()` is called.
`Scope.NOOP` is already provided by Brave. However, it has a different meaning. 
If we specify `Scope.NOOP`, `currentTraceContext.decorateScope(span.context(), Scope.NOOP)` really performs nothing.

Reply:
I see. Could you add some comment about it in the code? It'd be nice if Brave has a dedicated API for populating stuff without decorating, though. Did you have a chance to talk to @codefromthecrypt about this?

Reply:
Or is this intentional by design so the thread local variable is cleaned up when the decorated scope is closed?

Reply:
push/pop is messy vs allowing to stash state into the request per what rag said above. The impact of tricking out brave so that a no-op isn't a no-op is that you are guaranteeing overhead on each request. For example, it is more typical to use maybeScope which checks for no-op.

a single request In edge case could fork multiple context concurrently, etc, but that is horrifying and abnormal to do in a single span.

The thing about this change is that it is hacking to try to make the scope based api a push/pop which it isn't designed for. Any way you do this would be more messy than making a low-level set/get api in brave. However, the latter may lack the horsepower or will to land at all.

If things go this way, I suspect the best you can do is ack that there will be high overhead as MDC syncs are very expensive and the corresponding scope decorator in brave works to avoid those, or at least make them less. Covering with an object will guarantee any optimizations will not work.

IOTW it feels rock and a hard place.. the code is working around lack of an "unsafe context api" in brave. Even if we can't do that, perhaps we can take advantage of that `decorateScope` is protected and force this sort of more expensive behavior into the Armeria subclass.. that way it could be re-optimized. I feel like an approach on this path will be less trouble.

Reply:
one rule of thumb for instrumentation is "explicits are for you, implicit are for third party" in other words, only if there's no other way (and rarely this is the case when you control the code) always instrument your spans explicitly (ex context variables and explicit scope control.. aka no push/pop). "implicit" current trace context etc is for third party code, and in some cases it won't work well if people are managing their own threads, serializing to in-memory queues etc. Having a mental model which says what can be dodgy or not.. always choose not dodgy for span completion in instrumentation. If a span misses in 3rd party code it is less dire than your own code missing.

hope this helps!

Reply:
Thanks for the detail, @codefromthecrypt!

Reply:
Thanks, @codefromthecrypt! I know scope decorators are rather expensive and will activate it only when users specify `ScopeDecorator`s.

Reply:
sweet just know that may be always in the case of log mdc...

Reply:
ex there's a fair amount of logic internal to the scope decorators, which may help though.. you can add benchmarks similarly if it is easy to see impact when something like logging is on

https://github.com/openzipkin/brave/blob/master/context/slf4j/src/main/java/brave/context/slf4j/MDCScopeDecorator.java

https://github.com/openzipkin/brave/blob/master/instrumentation/benchmarks/src/main/java/brave/propagation/CurrentTraceContextBenchmarks.java

while it is costly in time to look at the benchmarks, it might help. ack that it might be too expensive and maybe just docs saying those looking for performance should do 'x' like disable the push/pop stuff

Reply:
I need some benchmarks when scope decorators are enabled. I wish it would not be too expensive. 🙏

Reply:
Fortunately, It seems like there is no significant performance regression.
TPS is decreased by 1% for simple GET requests with `BraveService` though results may vary depending on workloads.

ScopeDecorator|Requests/sec|Transfer/sec|Avg Req Time|Fastest Request|Slowest Request|Number of Errors
--------------|------------|------------|------------|---------------|---------------|----------------
MDC|39498.95|5.01MB|253.171µs|65.766µs|137.276202ms|0
MDC|40285.50|5.11MB|248.228µs|63.717µs|117.545369ms|0
MDC|40003.63|5.07MB|249.977µs|60.055µs|125.743473ms|0
MDC|40233.23|5.10MB|248.55µs|65.02µs|113.395833ms|0
Average of MDC|40005.3275|||||
No|40429.79|5.13MB|247.342µs|64.615µs|129.793452ms|0
No|40646.23|5.16MB|246.025µs|61.895µs|148.781964ms|0
No|40592.26|5.15MB|246.352µs|63.934µs|133.812856ms|0
No|40179.40|5.10MB|248.883µs|63.869µs|129.159041ms|0
No|40171.69|5.10MB|248.931µs|63.374µs|130.644515ms|0
Average of No|40403.874|||||